### PR TITLE
LIMIT is no column but a SQL keyword, allow limit on initial sync

### LINKED ISSUE #9339: Internal server error on REPORT DAV:sync-collection with result limit
### Steps to reproduce
1. Create an address book in Nextcloud and make sure it's available over CardDAV
2. Send a [REPORT DAV:sync-collection](https://tools.ietf.org/html/rfc6578#section-3.2) request with a body like this:

```
<?xml version='1.0' encoding='UTF-8' ?>
<sync-collection xmlns="DAV:">
  <sync-token>http://sabre.io/ns/sync/5</sync-token>
  <sync-level>1</sync-level>
  <limit><nresults>100</nresults></limit>
  <prop><getetag /></prop>
</sync-collection>
```

Note the `<limit><nresults>100</nresults></limit>` part which requests max. 100 results.

3. Nextcloud returns this 500 Internal server error:

```
    <d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
      <s:exception>Doctrine\DBAL\Exception\SyntaxErrorException</s:exception>
      <s:message>An exception occurred while executing 'SELECT `uri`, `operation` FROM `oc_addressbookchanges` WHERE `synctoken` &gt;= ? AND `synctoken` &lt; ? AND `addressbookid` = ? ORDER BY `synctoken` `LIMIT` 100' with params ["5", "8", "6"]:
    
    SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '`LIMIT` 100' at line 1</s:message>
    </d:error>
```

Also, when requesting an initial sync (no sync-token) with limit nresults=100, the server sends all results without batching. Instead, it should send 507.

### Expected behaviour

Nextcloud should return either up to 100 results, or, if this is not supported, HTTP 507 and an appropriate postcondition error code, see [RFC 6578 3.12 Example: DAV:sync-collection Report with Unsupported Limit](https://tools.ietf.org/html/rfc6578#section-3.12)

### Server configuration

**Nextcloud version:** 13.0.2

**Signing status:** No errors have been found.

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
The process control (PCNTL) extensions are required in case you want to interrupt long running commands - see http://php.net/manual/en/book.pcntl.php
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nc.dev001.net"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/nc.dev001.net",
        "dbtype": "mysql",
        "version": "13.0.2.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "maintenance": false,
        "trashbin_retention_obligation": "auto, 30",
        "theme": "",
        "loglevel": 2,
        "apps_paths": [
            {
                "path": "\/usr\/local\/www\/nextcloud\/apps",
                "url": "\/apps",
                "writable": true
            },
            {
                "path": "\/usr\/local\/www\/nextcloud\/apps-pkg",
                "url": "\/apps-pkg",
                "writable": false
            }
        ]
    }
}
```
</details>

**Are you using encryption:** no

### PR DESCRIPTION
**Pull request by georgehrke**
fixes #9339 

please test @rfc2822

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]
Should really move to the query builder at some time

**Comment by georgehrke** [comment]
I will take out the additional limit code and put it into a dedicated pull-request because it needs more testing ...

**Comment by rfc2822** [comment]
The first part (removing the `s) is working here fine :)

---

I'd not use the other part (allow limit on initial sync) this way because it returns the wrong sync-token (`$currentToken` instead of the sync-token that matches the sync state after applying the `limit nresults`) and thus breaks synchronization (which doesn't happen when the limit is silently ignored).

Support for client-requested limiting (truncating) is not required by the RFC, so maybe just throwing 507 on initial syncs with `limit nresults` would be a good idea (like https://tools.ietf.org/html/rfc6578#section-3.12)?

**Comment by blizzz** [comment]
set the milestone to 15 as this PR is against master.

**Comment by rullzer** [comment]
Moved it to the query builder (just so we keep cleaning up bit by bit).
@georgehrke please have a look as well.

If all is green we just need to rebase.

**Review by nickvergessen — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 1]
```diff
@@ -829,11 +835,19 @@ function getChangesForAddressBook($addressBookId, $syncToken, $syncLevel, $limit
...
-			$query = "SELECT `uri` FROM `*PREFIX*cards` WHERE `addressbookid` = ?";
-			$stmt = $this->db->prepare($query);
-			$stmt->execute([$addressBookId]);
+			$qb = $this->db->getQueryBuilder();
+			$qb->select('uri')
+				->from('cards')
+				->where($qb->expr()->eq('addressbookid', $qb->createNamedParameter($addressBookId)))
+				->orderBy('synctoken');
+
+			if ($limit>0) {
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/CardDavBackend.php]: Nitpick: add spaces around operator ;)

**Comment by rullzer** [comment]
CI is not happy. Seems that @georgehrke added the order by which is not present.

**Comment by MorrisJobke** [comment]
Moved to 15.0.1

**Comment by MorrisJobke** [comment]
conflicts 🙈

**Comment by georgehrke** [comment]
will look into it, but too late for 15.0.1

**Comment by rullzer** [comment]
Ah we should not sort by synctoken in tables that do not have one.

@georgehrke please have another look.

**Review by nickvergessen — APPROVED** [review]

**Comment by MorrisJobke** [comment]
What is the status here? Rebase and then ready for review?

**Comment by georgehrke** [comment]
Needs some work, on it ...

**Comment by MorrisJobke** [comment]
> Needs some work, on it ...

@georgehrke What is the status? We are close to the beta 1. 17 or 18?

**Comment by georgehrke** [comment]
> @georgehrke What is the status? We are close to the beta 1. 17 or 18?

Let me put it onto my ToDo list for tomorrow.

**Comment by MorrisJobke** [comment]
> Let me put it onto my ToDo list for tomorrow.

Tomorrow was yesterday, right? 😉

**Comment by georgehrke** [comment]
Test fresh sync with limit:

```
curl -X REPORT -u admin:batQz-Nc6zK-din8a-pdXoC-QPeWH -d '<?xml version="1.0" encoding="utf-8" ?>
   <D:sync-collection xmlns:D="DAV:">
     <D:sync-token/>
     <D:sync-level>1</D:sync-level>
     <D:limit>
       <D:nresults>1</D:nresults>
     </D:limit>
     <D:prop>
       <D:getetag/>
     </D:prop>
   </D:sync-collection>' 'http://nextcloud.local/remote.php/dav/calendars/admin/personal/'
```

=> Throws an exception 

Test fresh sync without limit:
```
curl -X REPORT -u admin:batQz-Nc6zK-din8a-pdXoC-QPeWH -d '<?xml version="1.0" encoding="utf-8" ?>
   <D:sync-collection xmlns:D="DAV:">
     <D:sync-token/>
     <D:sync-level>1</D:sync-level>
     <D:prop>
       <D:getetag/>
     </D:prop>
   </D:sync-collection>' 'http://nextcloud.local/remote.php/dav/calendars/admin/personal/'
```

=> Returns a list of all events

Test with sync-token and limit:
```
curl -X REPORT -u admin:batQz-Nc6zK-din8a-pdXoC-QPeWH -d '<?xml version="1.0" encoding="utf-8" ?>
   <D:sync-collection xmlns:D="DAV:">
     <D:sync-token>http://sabre.io/ns/sync/55</D:sync-token>
     <D:sync-level>1</D:sync-level>
     <D:limit>
       <D:nresults>1</D:nresults>
     </D:limit>
     <D:prop>
       <D:getetag/>
     </D:prop>
   </D:sync-collection>' 'http://nextcloud.local/remote.php/dav/calendars/admin/personal/'
```

=> Behaviour didn't change compared to master

Test with sync-token and without limit:
```
curl -X REPORT -u admin:batQz-Nc6zK-din8a-pdXoC-QPeWH -d '<?xml version="1.0" encoding="utf-8" ?>
   <D:sync-collection xmlns:D="DAV:">
     <D:sync-token>http://sabre.io/ns/sync/55</D:sync-token>
     <D:sync-level>1</D:sync-level>
     <D:limit>
       <D:nresults>1</D:nresults>
     </D:limit>
     <D:prop>
       <D:getetag/>
     </D:prop>
   </D:sync-collection>' 'http://nextcloud.local/remote.php/dav/calendars/admin/personal/'
```

=> Behaviour didn't change compared to master

**Comment by georgehrke** [comment]
I changed this PR as per @rfc2822's suggestion.
When a client sends an initial sync-request with a limit, we respond with a `{DAV:}:number-of-matches-within-limits` error

**Review by MorrisJobke — APPROVED** [review]
Code makes sense 👍

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**